include/kllvm/util/temporary_file.h

```diff
@@ -0,0 +1,55 @@
+#include <cstdlib>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <unistd.h>
+struct deleter {
+  void operator()(FILE *fp) const { std::fclose(fp); }
+};
+class temporary_file {
+private:
+  int temp_fd;
+  std::string temp_filename;
+  std::unique_ptr<FILE, deleter> temp_c_file;
+  std::optional<std::ofstream> temp_cpp_file;
+
+public:
+  temporary_file(std::string template_name)
+      : temp_fd(mkstemp(template_name.data()))
+      , temp_filename(template_name) {
+    if (temp_fd == -1) {
+      std::runtime_error("Could not create temporary file!");
+    }
+  }
+
+  ~temporary_file() {
+    close(temp_fd);
+    remove(temp_filename.data());
+  }
+
+  int descriptor() const { return temp_fd; }
+
+  std::string const &filename() const { return temp_filename; }
+
+  FILE *file_pointer(std::string const &mode = "r") {
+    if (!temp_c_file.get()) {
+      auto f = fdopen(temp_fd, mode.data());
+      if (f) {
+        temp_c_file = std::unique_ptr<FILE, deleter>(f);
+      } else {
+        std::runtime_error("Could not open file " + temp_filename);
```

Comment:
We can add the information from `perror` here; if a call that sets `errno` fails I almost always find myself looking for that information if it's not printed.

Reply:
```
} else {
        auto str = std::stringstream{};
        str << "Could not open file " << temp_filename;
        std::perror(str.str().c_str());
        std::runtime_error(str.str());
}
```
better this way?

Reply:
https://en.cppreference.com/w/cpp/string/byte/strerror

Sorry, my bad - I meant `std::strerror` rather than `perror`. That way we can put the error data into the exception, rather than forcing it to `stderr`.

Reply:
Done! I also notice that I wasn't throwing the errors and already fix it.